### PR TITLE
Fix the usage of pattern matching in module resolution for check of hasZeroOrOneAsteriskCharacter

### LINKED ISSUE #35171: TypeScript language service crashes because of wrong setting in tsconfig.json
*TS template added by @mjbvz*

**Confirmed TypeScript version**: 3.8.0-dev-20191116

**Search terms**:
- debug failure
- crash 

---

- VSCode Version: 1.40.1
- OS Version: macOS Mojave 10.14.6

While working on my TypeScript project I wanted to define something in `baseUrl` and `paths` fields of `compilerOptions` in `tsconfig.json` to avoid `../../../` hell. I did it for the first time so I did not do it correctly. Here's what I added:

```
    "baseUrl": "./src",               
    "paths": {
      "@interface/**/*" : ["./src/interface/*"],
      "@service/**/*": ["./src/service/**/*"],
      "@controller/**/*": ["./src/controller/**/*"],
    },  
```

after I did that, the built-in TypeScript language service crashed and did not want to start under no circumstances. It only started working again when I removed the new code from `tsconfig.json`

Steps to Reproduce:

1. Clone the drzamich/bot.ts@b4c97fb765be8e76cf6d6d8bb7a648e319545df8
2. Open the cloned project in VSCode and watch the TypeScript language service crash

Does this issue occur when all extensions are disabled?: Yes

### PR DESCRIPTION
**Pull request by sheetalkamat**
Fixes #35171

### CODE REVIEW

**Comment by sheetalkamat** [comment]
@RyanCavanaugh @DanielRosenwasser Do we want this for 3.7 as per https://github.com/microsoft/TypeScript/issues/35171#issuecomment-555177679

File: tests/baselines/reference/pathsValidation4.errors.txt [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+            "baseUrl": "./src",
+            "paths": {
+              "@interface/**/*" : ["./src/interface/*"],
+              ~~~~~~~~~~~~~~~~~
+!!! error TS5061: Pattern '@interface/**/*' can have at most one '*' character.
+              "@service/**/*": ["./src/service/**/*"],
+              ~~~~~~~~~~~~~~~
+!!! error TS5061: Pattern '@service/**/*' can have at most one '*' character.
+                                ~~~~~~~~~~~~~~~~~~~~
+!!! error TS5062: Substitution './src/service/**/*' in pattern '@service/**/*' in can have at most one '*' character.
```
Comment (DanielRosenwasser) [inline tests/baselines/reference/pathsValidation4.errors.txt]: What's with the "in can have" - is there an extra word there?

**Comment by DanielRosenwasser** [comment]
> Do we want this for 3.7

Is this a breaking change as well?

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-3.7 and LKG

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-3.7

File: tests/baselines/reference/pathsValidation4.errors.txt (hunk 1 above)
Comment (sheetalkamat) [reply tests/baselines/reference/pathsValidation4.errors.txt]: Fixing the existing error message..

**Comment by sheetalkamat** [comment]
> Is this a breaking change?

No but with vscode making updateOpen exceptions to close the tsserver it makes project with this configuration to crash LS completely

**Comment by sheetalkamat** [comment]
@typescript-bot cherry-pick this to release-3.7

**Pull request merged**